cpp/proto/substrait/extension_rels.proto

```diff
@@ -0,0 +1,32 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+syntax = "proto3";
+
+package arrow.substrait;
+
+import "substrait/algebra.proto";
+
+option csharp_namespace = "Arrow.Substrait";
+option go_package = "github.com/apache/arrow/substrait";
+option java_multiple_files = true;
+option java_package = "io.arrow.substrait";
+
+message AsOfJoinRel {
+  .substrait.Expression on = 1;
```

Comment:
This requires all the `on` column must have the same index for all tables?

Reply:
The proto requires a common expression, which could be a name rather than an index. OTOH, the Arrow Substrait code in this PR only deals with a field index expression; this can be extended, if desired.

Reply:
We are using `StructField` for this so I don't think we can use name for this, right?

I think it probably makes more sense to have this only be a repeated field in case that `on` column doesn't share the same column index for each table

Reply:
> We are using StructField for this so I don't think we can use name for this, right?

IIRC, Arrow Substrait currently only supports [`StructField` message](https://github.com/substrait-io/substrait/blob/f3f6bdc947e689e800279666ff33f118e42d2146/proto/substrait/algebra.proto#L856) but Substrait itself also provides a [`MapKey` message](https://github.com/substrait-io/substrait/blob/f3f6bdc947e689e800279666ff33f118e42d2146/proto/substrait/algebra.proto#L855) that can be used to reference by name.

> I think it probably makes more sense to have this only be a repeated field in case that on column doesn't share the same column index for each table

If we go down this path (see also [my comment here](https://github.com/apache/arrow/pull/14385#discussion_r994865351)), I think we'd also need to repeat the by-key, i.e., to wrap its existing repetition as a message and repeat that.